src/relax/transform/gradient.cc

```diff
@@ -0,0 +1,407 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file src/relax/transform/gradient.cc
+ * \brief A reverse-mode automatic differentiation.
+ *
+ * Now only supports differentiating a function in the IRModule with one dataflow block
+ * with respect to the only return value of the function. It needs to be scalar.
+ *
+ * For examples, see the MLP examples in tests/python/relax/test_transform_gradient.py and
+ * tests/python/relax/test_transform_gradient_numeric.py.
+ */
+
+#include <tvm/relax/expr_functor.h>
+#include <tvm/relax/op_attr_types.h>
+#include <tvm/relax/transform.h>
+
+#include <unordered_set>
+
+namespace tvm {
+namespace relax {
+
+class GradientMutator : public ExprMutator {
+ public:
+  explicit GradientMutator(IRModule mod, const Array<Var>& require_grads)
+      : ExprMutator(mod), require_grads_(require_grads) {}
```

Comment:
We can use `std::move` to avoid copying.
```suggestion
  explicit GradientMutator(IRModule mod, Array<Var> require_grads)
      : ExprMutator(std::move(mod)), require_grads_(std::move(require_grads)) {}
```

Reply:
Some questions here. 

It seems that `IRModule`, `Array`, `Var`, etc are `ObjectRef` objects that contains only one pointer. Its copy cost is very small. So what is the meaning of `std::move` and `const &` that used extensively in tvm (and specifically, in the pass infrastructure) ?

Reply:
And it seems that we do not have a move constructor for ObjectRef?

Reply:
The entire system is very large and it’s good to reduce copies. The move constructor is defined by default. https://github.com/mlc-ai/relax/blob/efefd723bf68e11da6b51baa92e983d5f6195b5b/include/tvm/runtime/object.h#L697-L705

Reply:
Oh I missed that definition... That makes sense!